packages/babel-plugin-minify-simplify/src/helpers.js

```diff
@@ -2,25 +2,8 @@
 
 const VOID_0 = t => t.unaryExpression("void", t.numericLiteral(0), true);
 
-// Types as Symbols - for comparing types
-// init must be empty object -
-// computing this involves checking object.keys() to be of length 0
-// skipped otherwise
-const types = {};
-const typeSymbols = t => {
-  // don't recompute
-  if (Object.keys(types).length < 1) {
-    t.TYPES.forEach(type => {
-      types[type] = Symbol.for(type);
-    });
-  }
-  return types;
-};
-
-const isNodeOfType = (t, node, typeSymbol) =>
```

Comment:
To me this function `isNodeOfType` looks like the reason for the slow-down. The `t["is" + Symbol.keyFor(typeForSymbol)]` property access looks particularly bad, since it's highly megamorphic and always passes a newly concatenated string, which means it'll always use the slow-path via the C++ runtime.

Reply:
@bmeurer Thanks for finding out the root cause.

Reply:
Did you verify my theory? I just guessed based on staring at your diff 😁

Reply:
My bad, I did have a look now and dont think that was the cause. It does have a minor increase though
## Before
<img width="442" alt="screen shot 2018-01-03 at 14 39 13" src="https://user-images.githubusercontent.com/3902525/34522668-0b93116e-f094-11e7-9d64-00ed8706eef6.png">

## After (Without polymorphic state)

<img width="450" alt="screen shot 2018-01-03 at 14 33 32" src="https://user-images.githubusercontent.com/3902525/34522669-0de787ec-f094-11e7-9c61-6074a552941a.png">

## Symbol creation logic
```js
const typeSymbols = t => {
  // don't recompute
  if (Object.keys(types).length < 1) {
    t.TYPES.forEach(type => {
      types[type] = Symbol.for(type);
    });
  }
  return types;
};
```

Still the problem of creating too many symbols lies as i stated before. 

<img width="397" alt="screen shot 2018-01-03 at 14 44 05" src="https://user-images.githubusercontent.com/3902525/34522779-9dfc4700-f094-11e7-9e5a-eefc008e452c.png">

Reply:
That's really odd. Can you run with `--runtime-call-stats` and search for `SymbolFor` in the output?

Reply:
Ok, so it's not `Symbol.for`. How often do you invoke `typeSymbols` function?

Reply:
We invoke twice per minification process. But we compute it only once and cache successive invocations.

Reply:
Can you post the first 10 items from `--runtime-call-stats`?

Reply:
Sure. 

```
        JS_Execution  22563.74ms  65.10%       780   0.01%
                  GC   5822.25ms  16.80%      6042   0.06%
    FunctionCallback   1567.68ms   4.52%      7327   0.07%
    KeyedGetProperty    742.17ms   2.14%   3932424  36.55%
RecompileSynchronous    729.72ms   2.11%      1414   0.01%
          ArrayConcat    401.79ms   1.16%    588038   5.47%
          ObjectKeys    389.82ms   1.12%     35271   0.33%
          StringSplit    379.40ms   1.09%   1751662  16.28%
    WeakCollectionSet    325.77ms   0.94%    523176   4.86%
          ArrayShift    136.54ms   0.39%    439691   4.09%
          ArraySlice    120.08ms   0.35%    278001   2.58%
          SetProperty    109.09ms   0.31%    272005   2.53%
ParseFunctionLiteral    104.88ms   0.30%      4990   0.05%


```

Reply:
The only relevant thing I can think of from the list is Object.keys which we are using as part of typeSymbols.

Reply:
After few confirmations, changing Object.keys with simple check is way faster. results here https://gist.github.com/vigneshshanmugam/c766550ecd02292dcdfbf0bf013b9d3d